lib/ui/window.dart

```diff
@@ -286,31 +569,212 @@ class Locale {
+      if (uAttributes != null && uAttributes.isNotEmpty)
+        result.write('-$uAttributes');
+      result.write(uOut.toString());
+    }
+    if (resultVWYZX.isNotEmpty)
+      result.write(resultVWYZX.toString());
+    if (extensions.containsKey('x')) {
+      result.write('-x');
+      if (extensions['x'].isNotEmpty) {
+        result.write('-${extensions["x"]}');
+      }
+    }
+    return result.toString();
+  }
+
+  // Unicode Language Identifier subtags
+  // TODO/WIP: because we lowercase Locale Identifiers before parsing, typical
+  // use of these regexps don't actually need to atch capitals too.
+  static final RegExp _reSingleton = RegExp(r'^[a-zA-Z]$');
+
+  // (https://www.unicode.org/reports/tr35/#Unicode_language_identifier).
+  static final RegExp _reLanguage = RegExp(r'^[a-zA-Z]{2,3}$|^[a-zA-Z]{5,8}$');
+  static final RegExp _reScript = RegExp(r'^[a-zA-Z]{4}$');
+  static final RegExp _reRegion = RegExp(r'^[a-zA-Z]{2}$|^[0-9]{3}$');
+  static final RegExp _reVariant = RegExp(r'^[a-zA-Z0-9]{5,8}$|^[0-9][a-zA-Z0-9]{3}$');
+  static final RegExp _reSep = RegExp(r'[-_]');
+
+  // Covers all subtags possible in Unicode Locale Identifiers, used for
+  // pu_extensions.
+  static final RegExp _reAllSubtags = RegExp(r'^[a-zA-Z0-9]{1,8}$');
+  // Covers all subtags within a particular extension, used for other_extensions.
+  static final RegExp _reOtherSubtags = RegExp(r'^[a-zA-Z0-9]{2,8}$');
+  // Covers "attribute" and "type" from unicode_locale_extensions, and "tvalue" in
+  // transformed_extensions.
+  // (https://www.unicode.org/reports/tr35/#Unicode_locale_identifier).
+  static final RegExp _reValueSubtags = RegExp(r'^[a-zA-Z0-9]{3,8}$');
+
+  static final RegExp _reKey = RegExp('^[a-zA-Z0-9][a-zA-Z]\$');
+  static final RegExp _reTkey = RegExp('^[a-zA-Z][0-9]\$');
+
   @override
   bool operator ==(dynamic other) {
     if (identical(this, other))
       return true;
     if (other is! Locale)
       return false;
     final Locale typedOther = other;
+
+    // Comparing Lists, Sets and Maps requires reimplementing functionality in
+    // package:collection. Variants and extensions are rare, so we convert them
+    // to canonical/normalized strings for comparison.
     return languageCode == typedOther.languageCode
-        && countryCode == typedOther.countryCode;
+        && countryCode == typedOther.countryCode
+        && scriptCode == typedOther.scriptCode
+        && variants.join('-') == typedOther.variants.join('-')
+        && _extensionsToString(_extensions) == _extensionsToString(typedOther._extensions);
   }
 
   @override
   int get hashCode {
     int result = 373;
     result = 37 * result + languageCode.hashCode;
+    if (_scriptCode != null)
+      result = 37 * result + scriptCode.hashCode;
     if (_countryCode != null)
       result = 37 * result + countryCode.hashCode;
+    if (_variants != null && _variants.isNotEmpty)
+      result = 37 * result + variants.join('-').hashCode;
```

Comment:
Maybe
```dart
  for (String variant in variants)
    result = 37 * result + variant.hashCode;
```
instead

Reply:
Switched to using "hashValues" from ui/hash_codes.dart, though this becomes a nuisance if we move this code out of Flutter/dart:ui.